crypto/rand/drbg_hmac.c

```diff
@@ -0,0 +1,172 @@
+}
+
+/* SP800-90Ar1 10.1.2.4 HMAC_DRBG_Rseed_Process: */
+static int drbg_hmac_reseed(RAND_DRBG *drbg,
+                            const unsigned char *ent, size_t ent_len,
+                            const unsigned char *adin, size_t adin_len)
+{
+    /* (Step 2) (K,V) = HMAC_DRBG_Update(entropy||additional_input, K, V) */
+    return drbg_hmac_update(drbg, ent, ent_len, adin, adin_len, NULL, 0);
+}
+
+/* SP800-90Ar1 10.1.2.5 HMAC_DRBG_Generate_Process: */
+static int drbg_hmac_generate(RAND_DRBG *drbg,
+                              unsigned char *out, size_t outlen,
+                              const unsigned char *adin, size_t adin_len)
+{
+    RAND_DRBG_HMAC *hmac = &drbg->data.hmac;
+    HMAC_CTX *ctx = hmac->ctx;
+    const unsigned char *temp = hmac->V;
+
+    if (adin != NULL
+            && adin_len > 0
+            && !drbg_hmac_update(drbg, adin, adin_len, NULL, 0, NULL, 0))
+        return 0;
+    for (;;) {
+        if (!HMAC_Init_ex(ctx, hmac->K, hmac->blocklen, hmac->md, NULL)
+                || !HMAC_Update(ctx, temp, hmac->blocklen))
+            return 0;
+
+        if (outlen > hmac->blocklen) {
+            if (!HMAC_Final(ctx, out, NULL))
+                return 0;
+            temp = out;
+        } else {
+            if (!HMAC_Final(ctx, hmac->V, NULL))
+                return 0;
+            memcpy(out, hmac->V, outlen);
+            break;
+        }
+        out += hmac->blocklen;
+        outlen -= hmac->blocklen;
+    }
+    if (!drbg_hmac_update(drbg, adin, adin_len, NULL, 0, NULL, 0))
+        return 0;
+
+    return 1;
+}
+
+static int drbg_hmac_uninstantiate(RAND_DRBG *drbg)
+{
+    HMAC_CTX_free(drbg->data.hmac.ctx);
+    OPENSSL_cleanse(&drbg->data.hmac, sizeof(drbg->data.hmac));
+    return 1;
+}
+
+static RAND_DRBG_METHOD drbg_hmac_meth = {
+    drbg_hmac_instantiate,
+    drbg_hmac_reseed,
+    drbg_hmac_generate,
+    drbg_hmac_uninstantiate
+};
+
+int drbg_hmac_init(RAND_DRBG *drbg)
+{
+    const EVP_MD *md = NULL;
+    RAND_DRBG_HMAC *hmac = &drbg->data.hmac;
+
+    /* Any approved digest is allowed - assume we pass digest (not NID_hmac*) */
+    md = EVP_get_digestbynid(drbg->type);
+    if (md == NULL)
+        return 0;
+
+    drbg->meth = &drbg_hmac_meth;
+
+    if (hmac->ctx == NULL) {
+        hmac->ctx = HMAC_CTX_new();
+        if (hmac->ctx == NULL)
+            return 0;
+    }
+
+    hmac->md = md;
+    hmac->blocklen = EVP_MD_size(md);
+    /* See SP800-57 Part1 Rev4 5.6.1 Table 3 */
+    drbg->strength = 64 * (hmac->blocklen >> 3);
+    if (drbg->strength > 256)
+        drbg->strength = 256;
+    drbg->seedlen = hmac->blocklen;
+
+    drbg->min_entropylen = drbg->strength / 8;
+    drbg->max_entropylen = DRBG_MINMAX_FACTOR * drbg->min_entropylen;
+
+    drbg->min_noncelen = drbg->min_entropylen / 2;
+    drbg->max_noncelen = DRBG_MINMAX_FACTOR * drbg->min_noncelen;
+
+    drbg->max_perslen = DRBG_MAX_LENGTH;
+    drbg->max_adinlen = DRBG_MAX_LENGTH;
+
+    drbg->max_request = 1 << 16;
```

Comment:
Where does this number come from?

Reply:
Added the following comments:
This one is a general one added up higher..
/* These are taken from SP 800-90 10.1 Table 2 */


/* Maximum number of bits per request = 2^19  = 2^16 bytes */